Problem with invalid Mathmode "AssertionError: Command \item invalid in math mode."

> import TexSoup
> latex_soup = TexSoup.TexSoup("""($3\times2\times$$\sim$5=$\sim$30 minutes each)""") 
> print(latex_soup)

I got the following "AssertionError: Command \item invalid in math mode." error in inline math equation. 
Please provide the solution for the above mentioned issue.

Thx for the colab notebook.

I played with it a bit, and I managed to make a minimal reproducible code snippet where the error occurs: 

```
texsrc = r"""
\documentclass[useAMS,usenatbib]{mnras}
\newcommand{\beq}{\begin{equation}}
\newcommand{\eeq}{\end{equation}}

\begin{document}
\begin{itemize}
\item something
\end{itemize}

\end{document}
"""
soup = TexSoup(texsrc)
```

RAISES: AssertionError: Command \item invalid in math mode.


In this situation, the shorthand macros `\beq` and `\eeq` screws up the parser for some reason, but I don't know why. I'm guessing the `\begin{equation}` in the macro definition puts us in math mode, and we never exit.

If you need to work on one manuscript specifically, I would recommend you remove the \beq and \eeq macro dfinitions would allow you to move forward... I just tried this deleting lines 58-61 in the file 1804.04667/main.tex and TexSoup was able to parse it

Thanks for figuring this out.In fact I will need to parse a couple manuscripts and not just this one. So having a better fix than deleting lines in the manuscript would help a lot.CheersTobiasAm 05.09.2023 um 00:50 schrieb Ivan Savov ***@***.***>:﻿
Thx for the colab notebook.
I played with it a bit, and I managed to make a minimal reproducible code snippet where the error occurs:
texsrc = r"""
\documentclass[useAMS,usenatbib]{mnras}
\newcommand{\beq}{\begin{equation}}
\newcommand{\eeq}{\end{equation}}

\begin{document}
\begin{itemize}
\item something
\end{itemize}

\end{document}
"""
soup = TexSoup(texsrc)

RAISES: AssertionError: Command \item invalid in math mode.
In this situation, the shorthand macros \beq and \eeq screws up the parser for some reason, but I don't know why. I'm guessing the \begin{equation} in the macro definition puts us in math mode, and we never exit.
If you need to work on one manuscript specifically, I would recommend you remove the \beq and \eeq macro dfinitions would allow you to move forward... I just tried this deleting lines 58-61 in the file 1804.04667/main.tex and TexSoup was able to parse it

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

--- a/TexSoup/reader.py
+++ b/TexSoup/reader.py
@@ -9,6 +9,7 @@ from TexSoup.tokens import (
     tokenize,
     SKIP_ENV_NAMES,
     MATH_ENV_NAMES,
+    SPECIAL_COMMANDS,
 )
 import functools
 import string
@@ -17,6 +18,7 @@ import sys
 
 MODE_MATH = 'mode:math'
 MODE_NON_MATH = 'mode:non-math'
+MODE_SPECIAL = 'mode:special'
 MATH_SIMPLE_ENVS = (
     TexDisplayMathModeEnv,
     TexMathModeEnv,
@@ -104,7 +106,9 @@ def read_expr(src, skip_envs=(), tolerance=0, mode=MODE_NON_MATH):
             assert mode != MODE_MATH, r'Command \item invalid in math mode.'
             contents = read_item(src)
             expr = TexCmd(name, contents, args, position=c.position)
-        elif name == 'begin':
+        # if we are in "special" mode, we do not attempt to match the `\begin`
+        # and `\end`
+        elif name == 'begin' and mode != MODE_SPECIAL:
             assert args, 'Begin command must be followed by an env name.'
             expr = TexNamedEnv(
                 args[0].string, args=args[1:], position=c.position)
@@ -546,9 +550,16 @@ def read_command(buf, n_required_args=-1, n_optional_args=-1, skip=0,
         next(buf)
 
     name = next(buf)
+    # if the command is a special one (like `newcommand`), enter "special"
+    # mode, in which a single `\begin` or `\end` are allowed
+    if name.text in SPECIAL_COMMANDS:
+        mode = MODE_SPECIAL
     token = Token('', buf.position)
     if n_required_args < 0 and n_optional_args < 0:
         n_required_args, n_optional_args = SIGNATURES.get(name, (-1, -1))
     args = read_args(buf, n_required_args, n_optional_args,
                      tolerance=tolerance, mode=mode)
+    # after parsing the command, go back to normal mode
+    if name.text in SPECIAL_COMMANDS:
+        mode = MODE_NON_MATH
     return name, args

--- a/TexSoup/tokens.py
+++ b/TexSoup/tokens.py
@@ -18,6 +18,7 @@ MATH_ENV_NAMES = (
     'eqnarray*', 'equation', 'equation*', 'flalign', 'flalign*', 'gather',
     'gather*', 'math', 'multline', 'multline*', 'split'
 )
+SPECIAL_COMMANDS = {'newcommand', 'renewcommand', 'providecommand'}
 BRACKETS_DELIMITERS = {
     '(', ')', '<', '>', '[', ']', '{', '}', r'\{', r'\}', '.' '|', r'\langle',
     r'\rangle', r'\lfloor', r'\rfloor', r'\lceil', r'\rceil', r'\ulcorner',

--- a/tests/test_parser.py
+++ b/tests/test_parser.py
@@ -629,3 +629,22 @@ def test_tolerance_env_unclosed():
     \end{envb}""", tolerance=1)
     assert len(list(soup.enva.contents)) == 1
     assert soup.end
+
+def test_special_command():
+    """Test that we tolerate unclosed environments when in special mode."""
+    # source:
+    # https://github.com/alvinwan/TexSoup/issues/135#issuecomment-1705749106
+    texsrc = r"""
+    \documentclass[useAMS,usenatbib]{mnras}
+    \newcommand{\beq}{\begin{equation}}
+    \newcommand{\eeq}{\end{equation}}
+
+    \begin{document}
+    \begin{itemize}
+    \item something
+    \end{itemize}
+
+    \end{document}
+    """
+    soup = TexSoup(texsrc)
+    assert soup

